File: LinkedList/Java/LinkedList.java
class LinkedList {
  Node head;
  Node tail;
  private int length = 0;

  static class Node {
    int data;
    Node next;

    Node(int data) {
      this.data = data;
    }
  }

  void prepend(int value) {
    Node node = new Node(value);
    node.next = this.head;
    this.head = node;
    if (this.tail == null) {
      this.tail = node;
    }
    this.length++;
  }
  void append(int value) {
    this.length++;
    Node node = new Node(value);
    if (this.head == null) {
      this.head = node;
      this.tail = node;

      return;
    }
    node.next = null;
    this.tail.next = node;
    this.tail = node;
  }

  void appendByIteration(int value) {
    this.length++;
    Node node = this.head;
    while(node != null) {
      if(node.next == null) {
        Node newNode = new Node(value);
        node.next = newNode;
        this.tail = newNode;
        return;
      }
      node = node.next;
    }
  }

  void unshift() {
    if(this.head == null) {
      throw new NullPointerException("Cannot delete from empty linked list");
    }
    this.length--;
    this.head = this.head.next;
  }

  void pop() {
    if (this.head == null) {
      throw new NullPointerException("Cannot delete from empty linked list");
    }
    this.length--;
    Node node = this.head;
    while(node != null) {
      Node currentNode = node;
      Node nextNode = node.next.next;
      if(nextNode == null) {
        currentNode.next = null;
        this.tail = currentNode;
        return;
      }
      node = node.next;
    }
  }

  void pop(int index) {
    if (this.head == null) {
      throw new NullPointerException("Cannot delete from empty linked list");
    }
    Node node = this.head;
    int counter = 0;
    if (index == length) {
      this.pop();
      return;
    } else if (index == 1) {
      this.unshift();
      return;
    }
    while(node != null) {
      counter++;
      if (counter + 1 == index) {
        node.next = node.next.next;
        return;
      }
      node = node.next;
    }
  }

  void fromArray(int arr[]) {
    for(int i = 0; i < arr.length; i++) {
      this.append(arr[i]);
    }
    this.length = arr.length;
  }

  void print() {
    Node node = this.head;
    while(node != null) {
      System.out.println(node.data);
      node = node.next;
    }
  }

  int getSize() {
    return this.length;
  }
  public static void main(String[] args) {
    LinkedList linkedList = new LinkedList();
    // Do operations here
  }
}

File: LinkedList/JavaScript/LinkedList.js
const LinkedListNode =  require('./LinkedListNode');

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }
  /* 
  First we have to create a new node. It's okay to pass a null head since we anyway need it null if the head is null
  Next, we say that this.head will be the node. Just in case if head was referenced to an already assigned node, then next of our new node
  will be the previous node and the head will now reference to the new node

  This solves the case for head

  Now coming to handling the tail, if tail is null we point it to the new node else we do nothing.
  
  */
  prepend(value) {
    const node = new LinkedListNode(value, this.head);
    this.head = node;

    if (!this.tail) {
      this.tail = node;
    }
  }
  append(value) {
    const node = new LinkedListNode(value);
    if (!this.head) {
      this.head = node;
      this.tail = node;

      return;
    }
    this.tail.next = node;
    this.tail = node;
  }
  appendByIteration(value) {
    let n = this.head;
    while(!n) {
      if(!n.next) {
        const node = new Node(value);
        n.next = node;
        this.tail = node;
        return;
      }
    }
  }

  print() {
    let nodes = this.head;
    while(!nodes) {
      console.log(nodes.data);
      nodes = nodes.next;
    }
  }
}

const node = new LinkedList();

node.append(10);
node.append(4);
node.prepend(5);
node.appendByIteration(9);

node.print();